### ISSUE TITLE
 sqlmlutils fails on install

### ISSUE THREAD

**Issue by adifranco1** [MicrosoftDocs/sql-docs#3782]
ERROR: Could not build wheels for pymssql which use PEP 517 and cannot be installed directly.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: bb790fba-2d19-b1a7-3935-6ec5f9207fcc
* Version Independent ID: f350cf6b-0fa8-6d76-8433-f3f752e62ecf
* Content: [Install Python packages with pip - SQL Server Machine Learning Services](https://docs.microsoft.com/en-us/sql/advanced-analytics/package-management/install-additional-python-packages-on-sql-server?view=sql-server-2017#feedback)
* Content Source: [docs/advanced-analytics/package-management/install-additional-python-packages-on-sql-server.md](https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/advanced-analytics/package-management/install-additional-python-packages-on-sql-server.md)
* Product: **sql**
* Technology: **machine-learning**
* GitHub Login: @garyericson
* Microsoft Alias: **garye**

**Comment by garyericson** [comment]
@adifranco1 - Is this the text of an error message you received? What were you doing at the time?

**Comment by adifranco1** [comment]
i ran this command "pip install --upgrade-strategy only-if-needed c:\temp\sqlmlutils-0.7.2.zip" as specified on Microsoft's site https://docs.microsoft.com/en-us/sql/advanced-analytics/package-management/install-additional-python-packages-on-sql-server?view=sql-server-2017#feedback

We're using SQL Server 2017 Enterprise. The documentation says we should be using sqlmlutils to load additional python libraries. Although, I get errors trying to install sqlmlutils. Our end goal is to use fbprophet for forecasting time series data.

**Comment by garyericson** [comment]
@adifranco1 - Sorry you're having trouble. Yes, sqlmlutils is the best way to add packages. Is this the error message you receive: "ERROR: Could not build wheels for pymssql which use PEP 517 and cannot be installed directly." Could you show all the messages you get from pip when you run it?

**Comment by adifranco1** [comment]
that is the exact error message i get

**Comment by adifranco1** [comment]
ERROR: Command errored out with exit status 1:
   command: 'c:\users\adifranco\appdata\local\programs\python\python38-32\python.exe' 'c:\users\adifranco\appdata\local\programs\python\python38-32\lib\site-packages\pip\_vendor\pep517\_in_process.py' build_wheel 'C:\Users\ADIFRA~1\AppData\Local\Temp\tmpfi0b1_zm'
       cwd: C:\Users\adifranco\AppData\Local\Temp\pip-install-y7qck0b9\pymssql
  Complete output (42 lines):
  running bdist_wheel
  running build
  installing to build\bdist.win32\wheel
  running install
  C:\Users\adifranco\AppData\Local\Temp\pip-build-env-1czxvxj1\overlay\Lib\site-packages\setuptools\dist.py:46: DistDeprecationWarning: Do not call this function
    warnings.warn("Do not call this function", DistDeprecationWarning)
  Traceback (most recent call last):
    File "c:\users\adifranco\appdata\local\programs\python\python38-32\lib\site-packages\pip\_vendor\pep517\_in_process.py", line 257, in <module>
      main()
    File "c:\users\adifranco\appdata\local\programs\python\python38-32\lib\site-packages\pip\_vendor\pep517\_in_process.py", line 240, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "c:\users\adifranco\appdata\local\programs\python\python38-32\lib\site-packages\pip\_vendor\pep517\_in_process.py", line 181, in build_wheel
      return _build_backend().build_wheel(wheel_directory, config_settings,
    File "C:\Users\adifranco\AppData\Local\Temp\pip-build-env-1czxvxj1\overlay\Lib\site-packages\setuptools\build_meta.py", line 208, in build_wheel
      return self._build_with_temp_dir(['bdist_wheel'], '.whl',
    File "C:\Users\adifranco\AppData\Local\Temp\pip-build-env-1czxvxj1\overlay\Lib\site-packages\setuptools\build_meta.py", line 194, in _build_with_temp_dir
      self.run_setup()
    File "C:\Users\adifranco\AppData\Local\Temp\pip-build-env-1czxvxj1\overlay\Lib\site-packages\setuptools\build_meta.py", line 236, in run_setup
      super(_BuildMetaLegacyBackend,
    File "C:\Users\adifranco\AppData\Local\Temp\pip-build-env-1czxvxj1\overlay\Lib\site-packages\setuptools\build_meta.py", line 142, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 452, in <module>
      setup(
    File "C:\Users\adifranco\AppData\Local\Temp\pip-build-env-1czxvxj1\overlay\Lib\site-packages\setuptools\__init__.py", line 145, in setup
      return distutils.core.setup(**attrs)
    File "c:\users\adifranco\appdata\local\programs\python\python38-32\lib\distutils\core.py", line 148, in setup
      dist.run_commands()
    File "c:\users\adifranco\appdata\local\programs\python\python38-32\lib\distutils\dist.py", line 966, in run_commands
      self.run_command(cmd)
    File "c:\users\adifranco\appdata\local\programs\python\python38-32\lib\distutils\dist.py", line 985, in run_command
      cmd_obj.run()
    File "C:\Users\adifranco\AppData\Local\Temp\pip-build-env-1czxvxj1\overlay\Lib\site-packages\wheel\bdist_wheel.py", line 228, in run
      self.run_command('install')
    File "c:\users\adifranco\appdata\local\programs\python\python38-32\lib\distutils\cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "c:\users\adifranco\appdata\local\programs\python\python38-32\lib\distutils\dist.py", line 985, in run_command
      cmd_obj.run()
    File "setup.py", line 449, in run
      raise DeprecationWarning(msg)
  DeprecationWarning: The pymssql project has been discontinued.  To install the last working released version, use a
  version specifier like "pymssql<3.0".  For details and alternatives see:
  https://github.com/pymssql/pymssql/issues/668
  ----------------------------------------
  ERROR: Failed building wheel for pymssql
  Running setup.py clean for pymssql
Failed to build pymssql
ERROR: Could not build wheels for pymssql which use PEP 517 and cannot be installed directly

**Comment by garyericson** [comment]
@adifranco1 - Thanks for the text. I'm checking with the developers, hoping that they haven't all left for the holidays yet.

**Comment by garyericson** [comment]
@adifranco1 - Okay, I discovered the issue. The pymssql package was discontinued just last month (you can read more [here](https://github.com/pymssql/pymssql/issues/668)), so the latest (default) version can no longer be installed and the error message suggests installing an earlier version of the package. Until we get sqlmlutils updated, you can install pymssql separately yourself just before installing sqlmlutils, and that should work (I've tried it and it worked for me). 

So you would do this:

```
pip install "pymssql<3.0"
pip install --upgrade-strategy only-if-needed c:\temp\sqlmlutils-0.7.2.zip
```

**Comment by garyericson** [comment]
I've updated the documentation to add the pymssql install: [Install Python packages with sqlmlutils](https://docs.microsoft.com/en-us/sql/advanced-analytics/package-management/install-additional-python-packages-on-sql-server?view=sql-server-2017#install-sqlmlutils-on-the-client-computer)

#please-close

**Comment by adifranco1** [comment]
Hi Gary,
I posted a new issue. Installing pymssql 3.0 fails to install. Any help is appreciated.

Anthony

________________________________

**Comment by mattfortech** [comment]
Hello adifranco1, I was wondering if you were able to get fbprophet installed using sqlmlutils? I've been struggling for the last week to get this installed and now I am at the point where all dependencies are installed successfully. I am very happy, HOWEVER now when fbprophet proceeds to install - it fails with a very cryptic error (see attached photo and text below)? I don't know what to do next. I have tried almost everything.

> Traceback (most recent call last):

 

>  File "C:\Users\Administrator\AppData\Roaming\Python\Python36\site-packages\sqlmlutils\sqlqueryexecutor.py", line 80, in execute_query
>     while self._cursor.nextset():
> pyodbc.ProgrammingError: ('42000', '[42000] [Microsoft][ODBC SQL Server Driver][SQL Server]A \'Python\' script error occurred during execution of \'sp_execute_external_script\' with HRESULT 0x80004004. (39004) (SQLMoreResults); [42000] [Microsoft][ODBC SQL Server Driver][SQL Server]An external script error occurred: \n\r\nError in execution.  Check the output for more information.\r\nTraceback (most recent call last):\r\n  File "<string>", line 5, in <module>\r\n  File "C:\\ProgramData\\MSSQLSERVER2\\Temp-PY\\Appcontainer1\\00F66B4C-5F7B-42A7-8BCE-04BB739D4F6B\\sqlindb_0.py", line 60, in transform\r\n    assert package_exists_in_scope("fbprophet", "public") != False\r\nAssertionError\r\n\r\nSqlSatelliteCall error: Error in execution.  Check the output for more information. (39019); [42000] [Microsoft][ODBC SQL Server Driver][SQL Server]STDOUT message(s) from external script: \nSqlSatelliteCall function failed. Please see the console output for more information.\r\nTraceback (most recent call last):\r\n  File "C:\\Program Files\\Microsoft SQL Server\\MSSQL15.MSSQLSERVER2\\PYTHON_SERVICES\\lib\\site-packages\\revoscalepy\\computecontext\\RxInSqlServer.py", line 605, in rx_sql_satellite_call\r\n    rx_native_call("SqlSatelliteCall", params)\r\n  File "C:\\Program Files\\Microsoft SQL Server\\MSSQL15.MSSQLSERVER2\\PYTHON_SERVICES\\lib\\site-packages\\revoscalepy\\RxSerializable.py", line 375, in rx_native_call\r\n    ret = px_call(functionname, params)\r\nRuntimeError: revoscalepy function failed.\r (39005)')
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "C:\Users\Administrator\AppData\Roaming\Python\Python36\site-packages\sqlmlutils\packagemanagement\sqlpackagemanager.py", line 186, in _install_many
>     self._install_single(sqlexecutor, target_package_file, scope, True, out_file=out_file)
>   File "C:\Users\Administrator\AppData\Roaming\Python\Python36\site-packages\sqlmlutils\packagemanagement\sqlpackagemanager.py", line 206, in _install_single
>     sqlexecutor.execute(builder, out_file=out_file)
>   File "C:\Users\Administrator\AppData\Roaming\Python\Python36\site-packages\sqlmlutils\sqlqueryexecutor.py", line 41, in execute
>     return self.execute_query(builder.base_script, builder.params, out_file=out_file)
>   File "C:\Users\Administrator\AppData\Roaming\Python\Python36\site-packages\sqlmlutils\sqlqueryexecutor.py", line 94, in execute_query
>     raise RuntimeError("Error in SQL Execution: " + str(e))
> RuntimeError: Error in SQL Execution: ('42000', '[42000] [Microsoft][ODBC SQL Server Driver][SQL Server]A \'Python\' script error occurred during execution of \'sp_execute_external_script\' with HRESULT 0x80004004. (39004) (SQLMoreResults); [42000] [Microsoft][ODBC SQL Server Driver][SQL Server]An external script error occurred: \n\r\nError in execution.  Check the output for more information.\r\nTraceback (most recent call last):\r\n  File "<string>", line 5, in <module>\r\n  File "C:\\ProgramData\\MSSQLSERVER2\\Temp-PY\\Appcontainer1\\00F66B4C-5F7B-42A7-8BCE-04BB739D4F6B\\sqlindb_0.py", line 60, in transform\r\n    assert package_exists_in_scope("fbprophet", "public") != False\r\nAssertionError\r\n\r\nSqlSatelliteCall error: Error in execution.  Check the output for more information. (39019); [42000] [Microsoft][ODBC SQL Server Driver][SQL Server]STDOUT message(s) from external script: \nSqlSatelliteCall function failed. Please see the console output for more information.\r\nTraceback (most recent call last):\r\n  File "C:\\Program Files\\Microsoft SQL Server\\MSSQL15.MSSQLSERVER2\\PYTHON_SERVICES\\lib\\site-packages\\revoscalepy\\computecontext\\RxInSqlServer.py", line 605, in rx_sql_satellite_call\r\n    rx_native_call("SqlSatelliteCall", params)\r\n  File "C:\\Program Files\\Microsoft SQL Server\\MSSQL15.MSSQLSERVER2\\PYTHON_SERVICES\\lib\\site-packages\\revoscalepy\\RxSerializable.py", line 375, in rx_native_call\r\n    ret = px_call(functionname, params)\r\nRuntimeError: revoscalepy function failed.\r (39005)')
> 

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Administrator\AppData\Roaming\Python\Python36\site-packages\sqlmlutils\packagemanagement\sqlpackagemanager.py", line 72, in install
    self._install_from_pypi(package, upgrade, version, install_dependencies, scope, out_file=out_file)
  File "C:\Users\Administrator\AppData\Roaming\Python\Python36\site-packages\sqlmlutils\packagemanagement\sqlpackagemanager.py", line 151, in _install_from_pypi
    self._install_from_file(target_package_file, scope, upgrade, out_file=out_file)
  File "C:\Users\Administrator\AppData\Roaming\Python\Python36\site-packages\sqlmlutils\packagemanagement\sqlpackagemanager.py", line 173, in _install_from_file
    self._install_many(target_package_file, dependencies_to_install, scope, out_file=out_file)
  File "C:\Users\Administrator\AppData\Roaming\Python\Python36\site-packages\sqlmlutils\packagemanagement\sqlpackagemanager.py", line 190, in _install_many
    raise RuntimeError("Package installation failed, installed dependencies were rolled back.") from e
RuntimeError: Package installation failed, installed dependencies were rolled back.


![dependenciessuccess](https://user-images.githubusercontent.com/40217248/88460171-23685600-ce4f-11ea-847a-012d491c5bef.png)
![sqllog](https://user-images.githubusercontent.com/40217248/88460212-86f28380-ce4f-11ea-9c9c-b10c32c0722d.png)

I tried to get more information from the log in SQL Server (see image) but there's not much there either?

**Comment by adifranco1** [comment]
I did get it to install but not with sqlmlutils. I can write up my method of installation in the next day or so if you like.

I worked with Microsoft on this as well. There’s some other issues we ran into regarding ML Server that I can brief you on.

________________________________

**Comment by mattfortech** [comment]
Omg!! I would be ETERNALLY GRATEFUL.
I have had very little sleep as I keep thinking I am getting closer. 
I’ve already screwed up my default instance of SQL trying different things.
Whatever instructions you could provide would be GREATLY appreciated.
Thank you so much!!!

**Comment by adifranco1** [comment]
mattfortech,
In our environment, we wanted to have the newest versions of python which requires "binding" to ML Server. If you're not familiar with binding it basically means you're downloading ML Server (as opposed to using ML Services that comes with SQL Server) which will swap out your Python or R files with newer libraries.

https://docs.microsoft.com/en-us/sql/machine-learning/install/upgrade-r-and-python?view=sql-server-2017

I don't remember if binding was required to get fbprophet to work but it's nice to be able to update your python libraries more frequently than waiting on a SQL Server release.

After that we tried to install fbprophet with sqlmlutils and pip but both failed (pip fails to install fbprophet on pc as well). However, Anaconda works every time. Here's the steps you can follow on your SQL Server.


  1.  Run a command prompt as administrator
  2.  Navigate to M:\Program Files\Microsoft SQL Server\MSSQL14\PYTHON_SERVICES\Scripts
  3.  Run command “conda install -c conda-forge fbprophet”

I also wanted to share that in our environment, every time our IT department install a SQL Server patch or CU it breaks our ML Server. I haven't found any information on the Internet of anyone else having this problem so this might be unique to our environment. We had to work with Microsoft on the fix and its less than ideal. Here are the steps to fix it.

C:\>cd C:\Program Files\Microsoft\ML Server\Setup

C:\Program Files\Microsoft\ML Server\Setup>sqlbindr.exe /list
MSSQL14.MSSQLSERVER

C:\Program Files\Microsoft\ML Server\Setup>sqlbindr.exe /unbind MSSQL14.MSSQLSERVER
<output truncated>
C:\Program Files\Microsoft\ML Server\Setup>sqlbindr.exe /bind MSSQL14.MSSQLSERVER /python
<output truncated>

.... now you have to install fbprophet (again) and any other libraries you installed.

In my opinion, ML Server (or ML Services) is a freaking toy for playing around with and not an enterprise solution. We have worked hard to come up with a solution where Python code resides in ".py" files and is incorporated into our code repository and CI/CD pipelines. Our plans are to move to Azure as soon as we can so we can use Data Bricks.

Here's a sample of a stored procedure where we use fbprophet.

create procedure [temp].[RunForecast]
@FilePath varchar(50)
,@FileName varchar(100)
,@SubjectArea varchar(100)
,@YearsBack smallint
,@StartYear smallint
,@EndYear smallint
,@DataParams varchar(8000)
,@HolidayParams varchar(8000)

as
begin

declare  @ExternalScriptCode nvarchar(max)
,@SQLString nvarchar(500)
,@ParameterDefinitions nvarchar(500)
,@database varchar(128) = db_name()
,@StartOfCurrentYear date = dateadd(yy, datediff(yy, 0, getdate()), 0)
;

-- FilePath + FileName is the path on the server where we place our .py files.
-- We are simply reading contents of the Python file into a variable and passing it into sp_execute_external_script
set @SQLString = N'select @FileOut = BulkColumn from openrowset(bulk ''' + @FilePath + @FileName + ''', single_clob) as Contents;';
set @ParameterDefinitions = N'@FileOut nvarchar(max) output';
execute master.sys.sp_executesql @SQLString, @ParameterDefinitions, @FileOut = @ExternalScriptCode output;

execute sp_execute_external_script
@language = N'Python'
,@script = @ExternalScriptCode
,@params = N'@subjectarea varchar(100), @yearsback int, @startyear int, @endyear int, @server varchar(100), @database varchar(128), @dataparams varchar(8000), @holidayparams varchar(8000)'
,@subjectarea = @SubjectArea
,@yearsback = @YearsBack
,@startyear = @StartYear
,@endyear = @EndYear
,@server = @@servername
,@database = @database
,@dataparams = @DataParams
,@holidayparams = @HolidayParams

end

If you need clarification on anything let me know.

Anthony


________________________________

**Comment by mattfortech** [comment]
Hello adifranco1,
Thank you so much for taking the time to help me. I really appreciate it. I have been spinning my tires on this for so long. I am going to create another environment and attempt your binding approach.  I will keep you posted.
Thank you again!!

**Comment by mattfortech** [comment]
Adifranco1,

I don’t know where to begin, in terms of thanking you for all your help! This is incredible. I have been struggling with this for SOOOOO long, and quite frankly was scared to death that the only way for me to introduce libraries into my environment was through sqlmlutils. There’s just no way!

This explains it!

Just to verify, here’s my plan to introduce/implement “binding,” install and make fbprophet available:

1.	Install Machine Learning Server for Windows (https://docs.microsoft.com/en-us/machine-learning-server/install/machine-learning-server-windows-install)
2.	Follow your instructions to install fbprophet using Anaconda

a.	Run a command prompt as administrator 
b.	Navigate to M:\Program Files\Microsoft SQL Server\MSSQL14\PYTHON_SERVICES\Scripts 
c.	Run command “conda install -c conda-forge fbprophet” 

 

Does this look correct?

Thank you ever so much, again, for all your help. I am so incredibly grateful.

**Comment by adifranco1** [comment]
Yep, those are the steps. You might find that when you install fbprophet that it prompts you to update other libraries but that’s about it.

Good luck and let me know if you run into any issues.

Connect with me!

http://linkedin.com/in/anthonyrdifranco

________________________________

**Comment by mattfortech** [comment]
Hello Adifranco1,

I have finished the Machine Learning Server for Windows install. I am curious to know how it knows which instance to bind to, but it seemed to install fine. No errors.

When I ran your instructions for after the installation, I had some questions:  If the new ML Server installed to say “C:\Program Files\Microsoft\ML Server\PYTHON_SERVER” then why am I running the conda script under “C:\Program Files\Microsoft SQL Server\MSSQL15.MSSQLSERVER\PYTHON_SERVICES\”? That is the original location, isn’t it?

As always, thank you so much for your time!

**Comment by adifranco1** [comment]
I'm not sure. I only had one instance on my server and it prompted me to add a new install or add it to an existing instance.

I had the same confusion. I originally ran conda from the new location and it didn't work.

You should see it in the results if you run the script below.

EXECUTE sp_execute_external_script   @language = N'Python',   @script = N' import pkg_resources import pandas as pd installed_packages = pkg_resources.working_set installed_packages_list = sorted(["%s==%s" % (i.key, i.version) for i in installed_packages]) df = pd.DataFrame(installed_packages_list) OutputDataSet = df 'WITH RESULT SETS (( PackageVersion nvarchar (150) ))


https://docs.microsoft.com/en-us/sql/machine-learning/package-management/python-package-information?view=sql-server-ver15

[https://docs.microsoft.com/en-us/media/logos/logo-ms-social.png]<https://docs.microsoft.com/en-us/sql/machine-learning/package-management/python-package-information?view=sql-server-ver15>
Get Python package information - SQL Server Machine Learning Services | Microsoft Docs<https://docs.microsoft.com/en-us/sql/machine-learning/package-management/python-package-information?view=sql-server-ver15>
This article describes how to get information about installed Python packages, including versions and installation locations, on Machine Learning Services on SQL Server and on Big Data Clusters.Example Python scripts show you how to list package information such as installation path and version.
docs.microsoft.com




________________________________

**Comment by mattfortech** [comment]
This prompt you received when running the Machine Learning Server 9.4.7 install, correct? Because I never received that prompt. I have two instances, but I don’t mind installing to either one – I just never saw the prompt.

I was poking around on the web and found this: https://docs.microsoft.com/en-us/sql/machine-learning/install/upgrade-r-and-python?view=sql-server-2017

 

Did you have to run the “SqlBindR.exe” command line tool to bind?

 

Thank you again for your help (and thank you for those links – incredibly beneficial)!

**Comment by adifranco1** [comment]
Yep, I don’t remember if it was that exact version but that’s what I installed. I installed through the GUI the time. Subsequent times when our patching broke it I used the command prompt.

If you run the command prompt and pass -list you can see what instance it’s installed in and maybe even pick an instance to install to.

________________________________
